### PR TITLE
Cleanup preparing for plumbing semaphores and timepoints between streams/hal.

### PR DESCRIPTION
**Pull request by benvanik**

### CODE REVIEW

**Comment by benvanik** [comment]
(was debating whether buffer_view -> tensor should be called import or export, but I think it makes sense as is when interpreted as "importing from an external buffer view into a tensor program")

**Comment by ScottTodd** [comment]
> (was debating whether buffer_view -> tensor should be called import or export, but I think it makes sense as is when interpreted as "importing from an external buffer view into a tensor program")

Yeah, I could see it going either way :P "importing from a tensor into the hal dialect" would also make sense...

**Comment by benvanik** [comment]
That's what gave me pause - but then stream.tensor.import/export (which these lower to) would be flipped (so a hal.tensor.import would become a stream.tensor.export, which feels wrong). I think really this may need to live in an abi dialect (where import and export to/from the tensor program is clearer) or something but haven't wanted to do that until I get the semaphore stuff figured out.

File: iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 1]
```diff
@@ -58,55 +58,94 @@ def HAL_ExSubmitAndWaitOp : HAL_Op<"ex.submit_and_wait", [YieldPoint]> {
...
+    "getTiedResultOperandIndices",
+  ]>,
+  Util_ShapeAwareOp,
+]> {
+  let summary = [{imports a tensor from a HAL buffer view}];
+  let description = [{
+    Defines an import of an external HAL buffer view into a SSA-form tensor.
+    An optional semaphore timepoint can be specified indicating when the
+    buffer view is available for use. If no semaphore timepoint is provided it
+    is assumed the buffer view is immediately available.
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/IR/HALOps.td:74]: Comment isn't true at this commit? (Fine to add that behavior later, just checking that this isn't a copy/paste issue)

**Review by ScottTodd — APPROVED** [review]

**Comment by ScottTodd** [comment]
> That's what gave me pause - but then stream.tensor.import/export (which these lower to) would be flipped (so a hal.tensor.import would become a stream.tensor.export, which feels wrong). I think really this may need to live in an abi dialect (where import and export to/from the tensor program is clearer) or something but haven't wanted to do that until I get the semaphore stuff figured out.

Yeah, I'm wondering looking at this how much we need hal->stream->hal vs just having an ABI dialect as you say then going abi->stream->hal.

File: iree/compiler/Dialect/HAL/IR/HALOps.td (hunk 1 above)
Comment (benvanik) [reply iree/compiler/Dialect/HAL/IR/HALOps.td:74]: ah yeah, when stella said I should probably do device enumeration stuff first I deleted the lines adding the semaphores as operands below :P will leave in here to cover the intent of the op even if it's not all wired up yet

**Comment by benvanik** [comment]
yeah, I think that's where I'm leaning too - the tricky part is that you need to specify HAL types at the ABI boundary otherwise we end up with ABI_Semaphore and HAL_Semaphore and ABI_BufferView and HAL_BufferView and that's just nasty. Once you need the types from a dialect you have to register the whole thing, so having duplicated ops feels a bit weird. Future me problems :P

**Pull request merged**